### ISSUE TITLE
Compatibility with Gradle 8

### ISSUE THREAD

**Issue by bddckr** [gmazzo/android-jacoco-aggregated-demo#1]
Hi! Thanks a lot for this demo project (and the blog post)!

I'm trying to update this demo project to work with the latest versions of the various dependencies. While the latest Android Gradle Plugin version has deprecated a few things that the coverage plugin uses (something I'm looking at later), a real blocking issue is encountered once you update Gradle to 8.0 (run `./gradlew wrapper --gradle-version=8.0`): The configuration `allCodeCoverageReportClassDirectories` no longer exists.

gradle/gradle#22616 seems to have the change in it.

I believe (configurations and variants are pretty new to me still!) what we need to do is to use [JacocoReportAggregationPlugin](https://github.com/gradle/gradle/blob/0ff0f8a55f7a3665f9463a6eb173c50d8bb727d6/subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoReportAggregationPlugin.java)'s `aggregateCodeCoverageReportResults` instead of `allCodeCoverageReportClassDirectories`.

However, once I do that, it looks like the variant matching fails to select our `codeCoverageElements` because `aggregateCodeCoverageReportResults` no longer is a configuration specific for source directories and instead uses `artifactView`s for the various attribute-based matching of data (sources + classes + execution data). We can no longer simply add our own attribute to change what gets selected as there doesn't seem to be any way to adjust how those artifact views are configured. There's no API to "get that artifact view to add an attribute to it".

tl;dr: Updating the version catalog to the following **and updating Gradle to 8.0** requires changes to the coverage plugin. Any idea what those are?
```toml
[versions]
agp = "7.4.1"
androidx-lifecycle = "2.5.1"
androidx-navigation = "2.5.3"
kotlin = "1.8.10"

[libraries]
androidx-annotation = "androidx.annotation:annotation:1.5.0"
androidx-appcompat = "androidx.appcompat:appcompat:1.6.1"
androidx-constraintlayout = "androidx.constraintlayout:constraintlayout:2.1.4"
androidx-core = "androidx.core:core-ktx:1.9.0"
androidx-livedata = { module = "androidx.lifecycle:lifecycle-livedata-ktx", version.ref = "androidx-lifecycle" }
androidx-navigation-fragment = { module = "androidx.navigation:navigation-fragment-ktx", version.ref = "androidx-navigation" }
androidx-navigation-ui = { module = "androidx.navigation:navigation-ui-ktx", version.ref = "androidx-navigation" }
androidx-test-espresso = "androidx.test.espresso:espresso-core:3.5.1"
androidx-test-junit = "androidx.test.ext:junit:1.1.5"
androidx-viewmodel = { module = "androidx.lifecycle:lifecycle-viewmodel-ktx", version.ref = "androidx-lifecycle" }
google-material = "com.google.android.material:material:1.8.0"
junit = "junit:junit:4.13.2"
robolectric = "org.robolectric:robolectric:4.9.2"

[plugins]
android = { id = "com.android.application", version.ref = "agp" }
android-lib = { id = "com.android.library", version.ref = "agp" }
kotlin-android = { id = "org.jetbrains.kotlin.android", version.ref = "kotlin" }
kotlin-jvm = { id = "org.jetbrains.kotlin.jvm", version.ref = "kotlin" }
```

**Comment by gmazzo** [comment]
Thanks for the very detailed report. I need to research how to make it work again as we are using a similar setup at my company.

> tl;dr: Updating the version catalog to the following and updating Gradle to 8.0 requires changes to the coverage plugin. Any idea what those are?
Not exactly sure what you mean here. Just bumping to Gradle 8 makes the plugin incompatible. There are other breaking changes?

**Comment by bddckr** [comment]
Oh sorry. I meant that we need to adjust the plugin, i.e., `coverage.gradle.kts` and was asking whether anyone knows the required changes to make it work again. You are correct that simply bumping to Gradle 8 gives the issues, so hopefully, you can reproduce this as well as a first step.

**Comment by gmazzo** [comment]
Actually the bump to AGP to `7.4.1` is also required, as seems it's not compatible also with Gradle 8

**Comment by gmazzo** [comment]
I've drafted a refactor to make it more compatible:
https://github.com/gmazzo/android-jacoco-aggregated-demo/pull/2

At this time I'm a bit block, I think there is an issue on `withVariantReselection` API, because one thing is to ignore all attributes from the owner configuration, but `AttributeDisambiguationRule` seems not to be getting invoked working properly (or the resolution logic seems is different).

Still researching.

**Comment by gmazzo** [comment]
Ok, finally made it work again. I don't like the idea of overriding `classDirectories`, but usage of `withVariantReselection` that Gradle 8 does breaks so many things. Tried using `capabilities`, disambiguation rules, but it seems it's broken.

**Comment by bddckr** [comment]
Thanks a lot! I can confirm that this is working for my project as well.

**Comment by gmazzo** [comment]
Hi @bddckr! I just wanted to let you know I've updated the code. Now not only works on Gradle 7/8 but also I don't rely on accessing/hacking internals :)
https://github.com/gmazzo/android-jacoco-aggregated-demo/commit/6d74b9cf899d35da53ea04dea06e1857a53dbf61

**Comment by bddckr** [comment]
That's fantastic! Thanks for pinging me. (Too bad GitHub doesn't allow setting up notifications for _commits_ yet.)

In my copy of this logic, I have started to change everything to
* use Gradle's lazy/task avoidance APIs like `Provider`/`Property` whenever possible
* only use AGP's _public_ API (`com.android.tools.build:gradle-api`)

I'll change my copy for your latest changes today and report how it went. I aim to raise a PR to discuss once I have something usable with the above changes.

**Comment by gmazzo** [comment]
That will be great! Let me know 💪
